File: dynamic_programming/house_robber_III.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

# O(n) time
# O(n) space, worst when h = n.
class Solution:
    def rob(self, root: Optional[TreeNode]) -> int:
        def dfs(node):
            if not node:
                return (0, 0)
            rob_left, skip_left = dfs(node.left)
            rob_right, skip_right = dfs(node.right)
            rob = node.val + skip_left + skip_right # rob here → skip children
            skip = max(rob_left, skip_left) + max(rob_right, skip_right) # skip here → best of each child
            return (rob, skip)
        return max(dfs(root))

File: dynamic_programming/house_robber_III.cpp
class Solution {
public:
    int rob(TreeNode* root) {
        std::pair<int,int> res = dfs(root);
        return std::max(res.first, res.second);
    }
private:
    std::pair<int,int> dfs(TreeNode* node) {
        if (!node) return {0, 0};

        std::pair<int,int> left  = dfs(node->left);
        std::pair<int,int> right = dfs(node->right);

        // OR:
        // auto [rob_left, skip_left] = dfs(node->left);

        int rob  = node->val + left.second + right.second;
        int skip = std::max(left.first, left.second)
                 + std::max(right.first, right.second);

        return {rob, skip};
    }
};

